### PR TITLE
Add function to get collapsed index from mesh

### LINKED ISSUE #862: Unit.Domain.Mesh Test Needs to be Factored
# Bug reports:

- The test body is too long and should be factor into separate functions that are called from the single test.
- Doesn't clang-format properly. We should probably avoid using `SECTION` and just call a function.

### PR DESCRIPTION
**Pull request by nilsdeppe**
## Proposed changes

Adds a function to get the collapsed/storage index from the Mesh, which corresponds to the 1D index in the `DataVector`s.

Fixes #862 

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

**Review by kidder — DISMISSED** [review]

File: tests/Unit/DataStructures/Test_Index.cpp [hunk 1]
```diff
@@ -16,6 +16,20 @@
...
+namespace {
+template <size_t Dim>
+void test_collapsed_index(const Index<Dim> extents) noexcept {
+  for (IndexIterator<Dim> index_it(extents); index_it; ++index_it) {
+    CAPTURE(*index_it);
+    std::array<size_t, Dim> index;
+    for (size_t d = 0; d < Dim; ++d) {
+      index[d] = (*index_it)[d];
+    }
+    CHECK(index_it.collapsed_index() == extents.collapsed_index(index));
```
Comment (nilsvu) [inline tests/Unit/DataStructures/Test_Index.cpp]: Please add comment like "Compare to the collapsed_index that is computed independently by the IndexIterator during iteration" to make clear this test is not redundant

File: src/DataStructures/Index.hpp [hunk 2]
```diff
@@ -118,12 +141,75 @@ class Index {
 /// Index. Note that the first dimension of the Index varies fastest when
 /// computing the collapsed index.
 template <size_t N>
-size_t collapsed_index(const Index<N>& index, const Index<N>& extents) noexcept;
+size_t collapsed_index(const Index<N>& index,
+                       const Index<N>& extents) noexcept {
+  return extents.collapsed_index(index.indices_);
```
Comment (nilsvu) [inline src/DataStructures/Index.hpp]: Why are you adding the member function, instead of keeping only the free function? I feel like we should choose one, and I remember you are always in favour of free functions over member functions

File: src/DataStructures/Index.hpp [hunk 3]
```diff
@@ -110,6 +123,16 @@ class Index {
   const std::array<size_t, Dim>& indices() const noexcept { return indices_; }
 
  private:
+  template <size_t N>
+  // NOLINTNEXTLINE(readability-redundant-declaration)
+  friend size_t collapsed_index(const Index<N>& index,
```
Comment (nilsvu) [inline src/DataStructures/Index.hpp]: Why does this need to be a `friend`? Looks like it only calls the public member function and could call `indices()` instead of using the private attribute

File: src/DataStructures/Index.hpp [hunk 4]
```diff
@@ -77,6 +79,17 @@ class Index {
   }
   /// \endcond
 
+  // @{
+  /*!
+   * \brief Returns the 1-dimensional index corresponding to the `Dim`
+   * dimensional `index`.
```
Comment (nilsvu) [inline src/DataStructures/Index.hpp]: Add a comment about the choice of fastest-varying direction, perhaps referring to `IndexIterator`. Somewhere there should be a reason documented for this choice.

File: src/DataStructures/Index.hpp [hunk 5]
```diff
@@ -118,12 +141,75 @@ class Index {
...
+               << d << " with value " << gsl::at(index, d)
+               << " exceeds the number of grid "
+                  "points "
+               << gsl::at(indices_, d));
+  }
+}
+#endif
+
+template <>
+SPECTRE_ALWAYS_INLINE size_t
```
Comment (nilsvu) [inline src/DataStructures/Index.hpp]: Please add a comment why you prefer this implementation over the previous, so that it won't be accidentally changed back in the future. My guess is that you want it to `inline` better?

Comment (nilsvu) [reply src/DataStructures/Index.hpp]: Also, why do these need to be in the header? Is that also for inlining?

**Review by nilsvu — CHANGES_REQUESTED** [review]
Please add comments in the code about the choices you made here (see questions below), so that the implementation won't be accidentally changed back in the future.

File: src/Domain/Mesh.hpp [hunk 6]
```diff
@@ -123,6 +123,14 @@ class Mesh {
    */
   size_t number_of_grid_points() const noexcept { return extents_.product(); }
 
+  /*!
+   * \brief Returns the 1-dimensional index corresponding to the `Dim`
+   * dimensional `index`.
```
Comment (nilsvu) [inline src/Domain/Mesh.hpp:128]: Also add a comment or reference about the fastest-varying direction here

File: src/Domain/Mesh.hpp [hunk 7]
```diff
@@ -123,6 +123,14 @@ class Mesh {
...
+   */
+  size_t storage_index(const std::array<size_t, Dim>& index) const noexcept {
```
Comment (nilsvu) [inline src/Domain/Mesh.hpp]: Add an overload that takes an `Index<Dim>`? Why even have the `std::array` overload (same in `Index.hpp`)?

**Review by nilsvu — CHANGES_REQUESTED** [review]

File: src/Domain/Mesh.hpp (hunk 7 above)
Comment (nilsdeppe) [reply src/Domain/Mesh.hpp]: Partly because I'm still confused why `Index` doesn't inherit off of `std::array` and I wanted `storage_index({{a,b,c}})` to work, though that may work with `Index` too, I'll test.

File: src/DataStructures/Index.hpp (hunk 2 above)
Comment (nilsdeppe) [reply src/DataStructures/Index.hpp]: Mostly to make it easier to find, but that's not a great reason. I'll change it.

**Comment by nilsdeppe** [comment]
rebased and posted fixups

**Comment by kidder** [comment]
I am fine with the fixups @nilsleiffischer ?

File: tests/Unit/DataStructures/Test_Index.cpp [hunk 8]
```diff
@@ -16,6 +16,23 @@
...
+template <size_t Dim>
+void test_collapsed_index(const Index<Dim> extents) noexcept {
+  for (IndexIterator<Dim> index_it(extents); index_it; ++index_it) {
+    CAPTURE(*index_it);
+    Index<Dim> index;
+    for (size_t d = 0; d < Dim; ++d) {
+      index[d] = (*index_it)[d];
+    }
+    // Compare the collapsed index of the IndexIterator to the collapsed index
+    // compute by the member function to make sure the two implementations are
```
Comment (nilsvu) [inline tests/Unit/DataStructures/Test_Index.cpp]: `member function` -> `free function`?

File: tests/Unit/Domain/Test_DomainHelpers.cpp [hunk 9]
```diff
@@ -926,18 +926,6 @@ SPECTRE_TEST_CASE("Unit.Domain.DomainHelpers.VolumeCornerIterator",
   test_vci_3d();
 }
 
-// [[OutputRegex, Index out of range.]]
-[[noreturn]] SPECTRE_TEST_CASE(
-    "Unit.Domain.DomainHelpers.VolumeCornerIterator.Assert", "[Domain][Unit]") {
```
Comment (nilsvu) [inline tests/Unit/Domain/Test_DomainHelpers.cpp:931]: This change seems unrelated to the rest of the PR

File: src/Domain/Mesh.hpp [hunk 10]
```diff
@@ -123,6 +123,18 @@ class Mesh {
...
 
+  /*!
+   * \brief Returns the 1-dimensional index corresponding to the `Dim`
+   * dimensional `index`.
+   *
+   * The first dimension varies fastest.
+   *
+   * \see collapsed_index()
+   */
+  size_t storage_index(const Index<Dim>& index) const noexcept {
```
Comment (nilsvu) [inline src/Domain/Mesh.hpp:134]: Calling this member function would be equivalent to the existing `collapsed_index(index, mesh.extents())`, so why would you like to add it? If you want to have this wrapper, would a free function be better also here?

**Review by nilsvu — COMMENTED** [review]
While squashing the current fixups, please change the first commit message to reflect that you are not adding the `collapsed_index` function but rewriting it

File: tests/Unit/Domain/Test_DomainHelpers.cpp (hunk 9 above)
Comment (nilsdeppe) [reply tests/Unit/Domain/Test_DomainHelpers.cpp:931]: The test doesn't test anything new and actually tests an assertion in collapsed index instead of in the iterator itself. I need to do something with it, so I'm removing it.

File: src/Domain/Mesh.hpp (hunk 10 above)
Comment (nilsdeppe) [reply src/Domain/Mesh.hpp:134]: Because it's useful and makes life easier. It also makes it so people will actually be aware it exists. I don't want it as a free function because I don't think it makes sense as a free function. The computation is strictly tied to a specific mesh and is part of that object.

Comment (nilsdeppe) [reply src/Domain/Mesh.hpp:134]: The other thing I want to make clear here is that this is the index used to store the data on the grid (hence storage_index, not collapsed_index). For the `Index` class this is not necessarily the relation ship, it is just a unique 1d index.

**Comment by nilsdeppe** [comment]
@nilsleiffischer squashed and pushed. Thanks for the feedback :)

@gsb76 could you please look at this?

**Review by mar-celine — DISMISSED** [review]
Looks good. One question for my own understanding: why the move of collapsed_index from the cpp file to the hpp file? Is the reason for this tied to avoiding the loops in the newer version? Or is that independent of the move?

**Comment by nilsdeppe** [comment]
The reason for moving to the hpp is to get inlining of the function. It's a very small, lightweight function (especially once the loops are removed) and so I want it inlined.

**Comment by nilsvu** [comment]
Looks ready to merge for me, except for the commit message that is outdated and that I suggested to change while squashing [in the comment above](https://github.com/sxs-collaboration/spectre/pull/1380#pullrequestreview-207164459)

**Comment by nilsdeppe** [comment]
@nilsleiffischer I've updated the commit message. Sorry for missing that.

**Review by mar-celine — APPROVED** [review]

**Review by kidder — APPROVED** [review]

**Review by nilsvu — APPROVED** [review]
